### ISSUE TITLE
Large File Uploads Fail

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1873]
|<img alt="lennyg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| lennyg reported|
|-|-|
|Trac ID|trac#1873|
|Type|defect|
|Created|2006-06-27 20:56:47Z|

resource.PostableResource doesn't seem able to handle file uploads much larger than 9M.  At 10M, the render() method is never entered.  Here is a code snippet that can be used to reproduce the error:

```
import os
from twisted.internet import reactor
from twisted.web2 import server, http, resource, channel
from twisted.web2 import http_headers, responsecode
from twisted.web2 import iweb, stream

SAVEDIR = "/tmp"
READSIZE=8192

class UploadFile(resource.PostableResource):
    def render(self, ctx):
        request = iweb.IRequest(ctx)
        filename = request.files['filename'][0][0]
        file = request.files['filename'][0][2]

        filestream = stream.FileStream(file)
        dest = os.path.join(SAVEDIR,filename)
        destfile = os.fdopen(os.open(dest,
                os.O_WRONLY | os.O_CREAT | os.O_EXCL,
                0644), 'w', 0)
        stream.readIntoFile(filestream, destfile)

        msg = "saved %s to %s" % (filename, dest)
        print msg
        return http.Response(stream="%s" % msg)

class Toplevel(resource.Resource):
    addSlash = True
    def render(self, ctx):
        return http.Response(responsecode.OK,
                {'content-type': http_headers.MimeType('text', 'html')},
                "Hello")

    child_uploadfile = UploadFile()

if __name__ == "__main__":
    site = server.Site(Toplevel())
    reactor.listenTCP(1080, channel.HTTPFactory(site))
    reactor.run()


And here is an html form that can be used to test, with files of varying sizes:

<html>
    <form action="http://localhost:1080/uploadfile"
            enctype="multipart/form-data" method="post">
        filename: <input type="file" name="filename">
        <input type="submit" value="submit">
    </form>
</html>
```

Programmatic clients (written in e.g. httplib) also get blocked upon writing only a few hunder KB if the file is larger than 10MB.

Attachments:

* [uploadserver.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/363/363078bca524b616d7944784f13556d3312630ef/c5f04ac2d632a679d7451b753818f1ce492ac964.py) (1107 bytes) - added by lennyg on 2006-06-27 20:58:12Z - uploatserver.py
* [uploadpost.html](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/363/363078bca524b616d7944784f13556d3312630ef/3c9fe5491e1b0b8a4d7d2a6f9cc23e25e8e20ddf.html) (209 bytes) - added by lennyg on 2006-06-27 20:59:10Z - html form for upload to uploadserver.py

<details><summary>Searchable metadata</summary>

```
trac-id__1873 1873
type__defect defect
reporter__lennyg lennyg
priority__normal normal
milestone__Web2_Gold_Master Web2-Gold-Master
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web2 web2
keywords__ 
time__1151441807000000 1151441807000000
changetime__1297608861000000 1297608861000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Modified submission for proper formatting.

**Comment by twisted-trac** [comment]
|<img alt="lennyg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lennyg commented|
|-|-|

Sorry -- didn't realize I could "Attach" (button didn't show up on entry screen) nor that inlined code would get all wonky.

**Comment by twisted-trac** [comment]
|<img alt="lennyg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|lennyg commented|
|-|-|

James Y. Knight wrote (http://twistedmatrix.com/pipermail/twisted-web/2006-June/002733.html):

The failing with > 10MB is intentional, but the hanging is not. It's  
supposed to return an error response to the client. In fileupload.py: 
233, parseMultipartFormData has arguments which limit the number of  
fields and the size of the uploaded data. The default limit for size  
is 10MB.

Unfortunately, it looks like there is no way to pass in different  
limits to fileupload.parseMultipartFormData from where it's called in  
server.parsePOSTData.

So there's two bugs here:
1) hangs instead of returning a response.
2) limit is not configurable.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Initial diagnostics show that it is indeed returning an error response but that the client doesn't get it. Perhaps because it's still blockingly trying to upload the rest of the file. It seems as though the server has paused reading the input (thus causing the client to block) rather than ignoring the rest.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Yes, that is exactly it. The server is supposed to close the connection when responding to a request before the complete request has been read, which it currently doesn't do.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @jyknight|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="lennyg's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|lennyg commented|
|-|-|

Replying to [therve](#note_8):

the previously included code isn't right.  the following demonstrates the problem properly:

```
#!/usr/bin/python

import os
from twisted.internet import reactor
from twisted.web2 import server, http, resource, channel, stream
from twisted.web2 import http_headers, responsecode
from twisted.web2 import iweb

SAVEDIR = "/tmp"

class UploadFile(resource.PostableResource):
    def render(self, ctx):
        request = iweb.IRequest(ctx)
        filename = request.files['filename'][0][0]
        file = request.files['filename'][0][2]

        filestream = stream.FileStream(file)
        dest = os.path.join(SAVEDIR,filename)
        destfile = os.fdopen(os.open(dest,
            os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0644), 'w', 0)
        d = stream.readIntoFile(filestream, destfile)
        d.addBoth(self.cleanup, destfile, file,
                "saving %s to %s" % (filename, dest))
        return d

    def cleanup(self, r, destfile, file, msg):
        destfile.close()
        file.close()

        if r:
            print "error: %s" % r
            msg = "%s failed" % msg
        else:
            msg = "%s succeeded" % msg
        print msg
        return http.Response(stream="%s" % msg)

class Toplevel(resource.Resource):
    addSlash = True
    def render(self, ctx):
        return http.Response(responsecode.OK,
                {'content-type': http_headers.MimeType('text', 'html')},
                "Hello")

    child_uploadfile = UploadFile()

if __name__ == "__main__":
    site = server.Site(Toplevel())
    reactor.listenTCP(1080, channel.HTTPFactory(site))
    reactor.run()
```

I'm still getting a hang with files >= 10MB, using 0.2.0+svn20070403, but very glad to see this targetted for GoldMaster!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

There is some stuff in the large-fileupload-1873 branch. The configuration of the limit is somewhat done. But the real bug mentioned here is not, because I don't manage to create an unit test for it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

(In [21310]) Merge forward work from large-fileupload-1873

Refs [#2836](https://github.com/twisted/twisted/issues/2836)
Refs [#1873](https://github.com/twisted/twisted/issues/1873)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

Not going to fix this, see [#4821](https://github.com/twisted/twisted/issues/4821).  Twisted Web doesn't have a similar problem with large uploads, so not retargeting at the web component either.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
